File: sdks/python/src/splyntra/governance.py
# SPDX-License-Identifier: Apache-2.0
"""Governance helpers — call the collector's decision + ledger endpoints.

    from splyntra import authorize, log_action
    d = authorize("payments.refund", agent_id="support_agent", context={"amount": 80})
    if d["decision"] == "allow":
        ...  # proceed
    elif d["decision"] == "needs_approval":
        ...  # wait for a human

These call the collector (same endpoint/key as tracing).
"""

from __future__ import annotations

import json
import os
import urllib.request
from typing import Optional


def _endpoint() -> str:
    return os.getenv("SPLYNTRA_ENDPOINT", "http://localhost:4318").rstrip("/")


def _key(explicit: Optional[str]) -> str:
    key = explicit or os.getenv("SPLYNTRA_API_KEY", "")
    if not key:
        raise ValueError("Splyntra: set SPLYNTRA_API_KEY or pass api_key=")
    return key


def _post(path: str, payload: dict, api_key: Optional[str]) -> dict:
    req = urllib.request.Request(
        f"{_endpoint()}{path}",
        data=json.dumps(payload).encode("utf-8"),
        method="POST",
        headers={"Authorization": f"Bearer {_key(api_key)}", "Content-Type": "application/json"},
    )
    with urllib.request.urlopen(req, timeout=15) as resp:
        return json.loads(resp.read().decode("utf-8"))


def authorize(action: str, agent_id: str = "agent", resource: str = "", context: Optional[dict] = None, api_key: Optional[str] = None) -> dict:
    """Ask whether an agent may perform an action on a resource.

    ``resource`` scopes the decision (e.g. ``"payroll.read"``) so resource-level
    policies can match; omit it for action-only rules.

    Returns ``{"decision": "allow"|"deny"|"needs_approval", ...}``.
    """
    return _post(
        "/v1/authorize",
        {"agent_id": agent_id, "action": action, "resource": resource, "context": context or {}},
        api_key,
    )


def log_action(action: str, actor: str = "agent", resource: str = "", trace_id: str = "", metadata: Optional[dict] = None, api_key: Optional[str] = None) -> dict:
    """Append a consequential action to the immutable activity ledger."""
    return _post(
        "/v1/ledger",
        {"actor": actor, "action": action, "resource": resource, "trace_id": trace_id, "metadata": metadata or {}},
        api_key,
    )

File: sdks/typescript/src/governance.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Governance helpers — call the collector's decision + ledger endpoints (npm
 * parity with the Python `authorize` / `log_action`).
 *
 *   import { authorize, logAction } from "@splyntra/sdk";
 *   const d = await authorize("payments.refund", { agentId: "support", context: { amount: 80 } });
 *   if (d.decision === "allow") { ... }
 *   else if (d.decision === "needs_approval") { ... wait for a human ... }
 *
 * These hit the collector (same endpoint/key as tracing). The decision + ledger
 * routes are served by the commercial governance module.
 */
function endpoint(): string {
  return (process.env.SPLYNTRA_ENDPOINT || "http://localhost:4318").replace(/\/$/, "");
}
function apiKey(explicit?: string): string {
  const k = explicit || process.env.SPLYNTRA_API_KEY || "";
  if (!k) throw new Error("Splyntra: set SPLYNTRA_API_KEY or pass apiKey");
  return k;
}
async function post(path: string, payload: unknown, key?: string): Promise<any> {
  const res = await fetch(`${endpoint()}${path}`, {
    method: "POST",
    headers: { Authorization: `Bearer ${apiKey(key)}`, "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error(`governance request failed (${res.status}): ${await res.text().catch(() => "")}`);
  return res.json();
}

export interface AuthorizeDecision {
  decision: "allow" | "deny" | "needs_approval";
  reason?: string;
  request_id?: string;
  limit_usd?: number;
}

/** Ask whether an agent may perform an action on a resource. `resource` scopes
 * the decision (e.g. "payroll.read") for resource-level policies; omit it for
 * action-only rules. */
export async function authorize(
  action: string,
  opts: { agentId?: string; resource?: string; context?: Record<string, unknown>; apiKey?: string } = {}
): Promise<AuthorizeDecision> {
  return post("/v1/authorize", { agent_id: opts.agentId || "agent", action, resource: opts.resource || "", context: opts.context || {} }, opts.apiKey);
}

/** Append a consequential action to the immutable, tamper-evident activity ledger. */
export async function logAction(
  action: string,
  opts: { actor?: string; resource?: string; traceId?: string; metadata?: Record<string, unknown>; apiKey?: string } = {}
): Promise<any> {
  return post("/v1/ledger", { actor: opts.actor || "agent", action, resource: opts.resource || "", trace_id: opts.traceId || "", metadata: opts.metadata || {} }, opts.apiKey);
}
